src/wayland/idle_notify/mod.rs

```diff
@@ -0,0 +1,345 @@
+//! # }
+//! impl IdleNotifierHandler for State {
+//!     fn idle_notifier_state(&mut self) -> &mut IdleNotifierState<Self> {
+//!         &mut self.idle_notifier
+//!     }
+//! }
+//! delegate_idle_notify!(State);
+//!
+//! // On input you should notify the idle_notifier
+//! // state.idle_notifier.notify_activity(&wl_seat);
+//! ```
+
+use std::{
+    collections::HashMap,
+    sync::{
+        atomic::{self, AtomicBool},
+        Mutex,
+    },
+    time::Duration,
+};
+
+use calloop::{timer::TimeoutAction, LoopHandle, RegistrationToken};
+use wayland_protocols::ext::idle_notify::v1::server::{
+    ext_idle_notification_v1::{self, ExtIdleNotificationV1},
+    ext_idle_notifier_v1::{self, ExtIdleNotifierV1},
+};
+use wayland_server::{
+    backend::{ClientId, GlobalId, ObjectId},
+    protocol::wl_seat::WlSeat,
+    Client, DataInit, Dispatch, DisplayHandle, GlobalDispatch, New, Resource,
+};
+
+/// Handler trait for ext idle notify module
+pub trait IdleNotifierHandler: Sized {
+    /// [IdleNotifierSeatState] getter
+    fn idle_notifier_state(&mut self) -> &mut IdleNotifierState<Self>;
+}
+
+/// User data of [ExtIdleNotificationV1] resource
+#[derive(Debug)]
+pub struct IdleNotificationUserData {
+    seat: WlSeat,
+    is_idle: AtomicBool,
+    timeout: Duration,
+    timer_token: Mutex<Option<RegistrationToken>>,
+}
+
+impl IdleNotificationUserData {
+    fn take_timer_token(&self) -> Option<RegistrationToken> {
+        self.timer_token.lock().unwrap().take()
+    }
+
+    fn set_timer_token(&self, idle: Option<RegistrationToken>) {
+        *self.timer_token.lock().unwrap() = idle;
+    }
+
+    fn set_idle(&self, idle: bool) {
+        self.is_idle.store(idle, atomic::Ordering::Release);
+    }
+
+    fn is_idle(&self) -> bool {
+        self.is_idle.load(atomic::Ordering::Acquire)
+    }
+}
+
+type SeatObjectId = ObjectId;
+type NotificationObjectId = ObjectId;
+
+type SeatMap<T> = HashMap<SeatObjectId, T>;
+type NotificationMap<T> = HashMap<NotificationObjectId, T>;
+
+/// State of ext-idle-notify module
+#[derive(Debug)]
+pub struct IdleNotifierState<D> {
+    global: GlobalId,
+    notifications: SeatMap<NotificationMap<ExtIdleNotificationV1>>,
+    loop_handle: LoopHandle<'static, D>,
+    is_inhibited: bool,
+}
+
+impl<D: IdleNotifierHandler> IdleNotifierState<D> {
+    /// Create new [`ExtIdleNotifierV1`] global.
+    pub fn new(display: &DisplayHandle, loop_handle: LoopHandle<'static, D>) -> Self
+    where
+        D: GlobalDispatch<ExtIdleNotifierV1, ()>,
+        D: Dispatch<ExtIdleNotifierV1, ()>,
+        D: Dispatch<ExtIdleNotificationV1, IdleNotificationUserData>,
+        D: IdleNotifierHandler,
+        D: 'static,
+    {
+        let global = display.create_global::<D, ExtIdleNotifierV1, _>(1, ());
+        Self {
+            global,
+            notifications: HashMap::new(),
+            loop_handle,
+            is_inhibited: false,
+        }
+    }
```

Comment:
Anvil uses a different state for implementing states and for the event loop data, making this very inconvenient.

I'd assume that this is why you didn't bother with the Anvil implementation?

Reply:
> I'd assume that this is why you didn't bother with the Anvil implementation?

No, I did not consider that, but that's exactly why I wanted to do anvil impl before I drafting, to catch stuff like that. 

Sounds like this can be fixed with additional generic, will try to play around with that.